### PR TITLE
Add support for sticky (/y)

### PR DESCRIPTION
**Pull request by Exter-N**
This PR :
- Fixes the combination of the `^` operator and `/g` (see also #29) ;
- Refactors some parts of many of the algorithms of the library ;
- Adds support for `/y` (see also #23) ;
- Makes the library more exception-friendly (in many cases, if some code called back by the library throws an exception, it'll be catchable by the caller, instead of crashing Node).

### CODE REVIEW

File: lib/exec.cc [hunk 1]
```diff
@@ -25,50 +53,18 @@ NAN_METHOD(WrappedRE2::Exec) {
...
-		data = node::Buffer::Data(info[0]);
-	} else {
-		if (re2->global && re2->lastIndex) {
-			String::Value s(info[0]->ToString());
-			if (re2->lastIndex > s.length()) {
-				re2->lastIndex = 0;
-				info.GetReturnValue().SetNull();
-				return;
-			}
-			Local<String> t(Nan::New(*s + re2->lastIndex).ToLocalChecked());
```
Comment (Exter-N) [inline lib/exec.cc]: I'm looking for a way to check in which cases this optimization can safely be restored, and to do so when possible.

**Comment by Exter-N** [comment]
I've been able to fully restore the optimization when it looks safe to do so, and nearly fully even in other cases (as the engine doesn't have any lookaround capabilities by design, it actually only needs to know at most one character before the starting position) and therefore deleted my comment about this point.

That makes it OK for me now. What do you think?

**Comment by uhop** [comment]
Currently I am traveling for a conference. I’ll review the PR and messages when I am back in the beginning of the next week.

**Comment by GuilhemN** [comment]
Would be a great addition :+1:

**Comment by uhop** [comment]
It is definitely something I want to see in the library. But it is a major feature, and the pull request has restructured a lot of files, so I am going over them with a fine-tooth comb. Hang on.

**Pull request closed without merging**